app/include/zmk/event_manager.h

```diff
@@ -68,17 +71,22 @@ struct zmk_event_subscription {
             .listener = &zmk_listener_##mod,                                                       \
     };
 
-#define ZMK_EVENT_RAISE(ev) zmk_event_manager_raise((zmk_event_t *)ev);
+#define ZMK_EVENT_RAISE(ev)                                                                        \
+    ((__ASSERT((uint8_t *)&(ev).header - (uint8_t *)&ev == 0,                                      \
+               "header must be first element of event")),                                          \
+     zmk_event_manager_raise(&(ev).header));
 
 #define ZMK_EVENT_RAISE_AFTER(ev, mod)                                                             \
-    zmk_event_manager_raise_after((zmk_event_t *)ev, &zmk_listener_##mod);
+    ((__ASSERT((uint8_t *)&(ev).header - (uint8_t *)&ev == 0,                                      \
+               "header must be first element of event")),                                          \
+     zmk_event_manager_raise_after(&(ev).header, &zmk_listener_##mod));
 
 #define ZMK_EVENT_RAISE_AT(ev, mod)                                                                \
-    zmk_event_manager_raise_at((zmk_event_t *)ev, &zmk_listener_##mod);
-
-#define ZMK_EVENT_RELEASE(ev) zmk_event_manager_release((zmk_event_t *)ev);
+    ((__ASSERT((uint8_t *)&(ev).header - (uint8_t *)&ev == 0,                                      \
+               "header must be first element of event")),                                          \
+     zmk_event_manager_raise_at(&(ev).header, &zmk_listener_##mod));
 
-#define ZMK_EVENT_FREE(ev) k_free((void *)ev);
+#define ZMK_EVENT_RELEASE(ev) zmk_event_manager_release((zmk_event_t *)&ev);
```

Comment:
```suggestion
#define ZMK_EVENT_RELEASE(ev) zmk_event_manager_release((zmk_event_t *)&ev)
```

Also, I think we want to add the assert and use `&ev.header` instead of a cast here too?